flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterFactory.java

```diff
@@ -0,0 +1,174 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.codegeneration;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeutils.TypeComparator;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.runtime.operators.sort.FixedLengthRecordSorter;
+import org.apache.flink.runtime.operators.sort.InMemorySorter;
+import org.apache.flink.runtime.operators.sort.NormalizedKeySorter;
+
+import freemarker.template.TemplateException;
+
+import org.codehaus.commons.compiler.CompileException;
+import org.codehaus.janino.SimpleCompiler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * {@link SorterFactory} is a singleton class that provides functionalities to create the most suitable sorter
+ * for underlying data based on {@link TypeComparator}.
+ * Note: the generated code can be inspected by configuring Janino to write the code that is being compiled
+ * to a file, see http://janino-compiler.github.io/janino/#debugging
+ */
+public class SorterFactory {
+	// ------------------------------------------------------------------------
+	//                                   Constants
+	// ------------------------------------------------------------------------
+	private static final Logger LOG = LoggerFactory.getLogger(SorterFactory.class);
+
+	/** Fixed length records with a length below this threshold will be in-place sorted, if possible. */
+	private static final int THRESHOLD_FOR_IN_PLACE_SORTING = 32;
+
+	// ------------------------------------------------------------------------
+	//                                   Singleton Attribute
+	// ------------------------------------------------------------------------
+	private static SorterFactory sorterFactory;
+
+	// ------------------------------------------------------------------------
+	//                                   Attributes
+	// ------------------------------------------------------------------------
+	private SimpleCompiler classCompiler;
+	private TemplateManager templateManager;
+	private HashMap<String, Constructor> constructorCache;
+
+	/**
+	 * This is only for testing. If an error occurs, we want to fail the test, instead of falling back
+	 * to a non-generated sorter.
+	 */
+	public boolean forceCodeGeneration = false;
+
+	/**
+	 * Constructor.
+	 */
+	private SorterFactory() {
+		this.templateManager = TemplateManager.getInstance();
+		this.classCompiler = new SimpleCompiler();
```

Comment:
Thanks for raising this issue. Actually it's a bit more complicated than just setting the parent classloader, because of the caching of the generated classes. The problem is that the `constructorCache` would keep the old user code classloader alive. I'm currently thinking of the following solution:
1. include also the user code classloader in the key of the cache, and
2. use `WeakReference` for both the values and keys of the cache.

This will ensure that we don't try to reuse generated classes from a previous job, since the classloader will be different across jobs, so we will have different keys. And the cache won't keep anything alive, since it will only have `WeakReferences`.
(For 2., I will use the Guava `CacheBuilder`.)

Reply:
909b59e3174a5348c2ee550806d4711475f5c268
cce40c5cd4aaff446bb4bec8918d2fda37649e0a

This was a bit tricky, I hope I haven't messed it up. I will think about this a bit more in the next few days, and maybe do some more testing, to see that we are not keeping alive anything from past jobs. Unfortunately, I don't know how this could be tested in an automated way. I think I will test it manually by just submitting hundreds of jobs, and watching in a profiler that object counts are not growing.